File: MuscleHamsterExpo/src/components/NotificationContextBanner.js
// Notification Context Banner - Phase 08.3
// Contextual banner shown when app opens from a notification tap

import React, { useState, useEffect, useRef } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  Animated,
  AccessibilityInfo,
} from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { BannerTypeInfo } from '../models/Notification';

export default function NotificationContextBanner({
  context,
  onDismiss,
  onAction,
}) {
  const [isVisible, setIsVisible] = useState(false);
  const opacity = useRef(new Animated.Value(0)).current;
  const translateY = useRef(new Animated.Value(-20)).current;
  const dismissTimeoutRef = useRef(null);

  useEffect(() => {
    if (context) {
      animateIn();
    } else {
      setIsVisible(false);
    }

    return () => {
      if (dismissTimeoutRef.current) {
        clearTimeout(dismissTimeoutRef.current);
      }
    };
  }, [context]);

  const animateIn = () => {
    setIsVisible(true);

    Animated.parallel([
      Animated.spring(opacity, {
        toValue: 1,
        useNativeDriver: true,
      }),
      Animated.spring(translateY, {
        toValue: 0,
        friction: 8,
        tension: 40,
        useNativeDriver: true,
      }),
    ]).start();

    // Announce for accessibility
    setTimeout(() => {
      AccessibilityInfo.announceForAccessibility(context.bannerMessage);
    }, 500);

    // Schedule auto-dismiss if applicable
    if (context.shouldAutoDismiss) {
      dismissTimeoutRef.current = setTimeout(() => {
        handleDismiss();
      }, context.autoDismissDelay * 1000);
    }
  };

  const handleDismiss = () => {
    if (dismissTimeoutRef.current) {
      clearTimeout(dismissTimeoutRef.current);
    }

    Animated.parallel([
      Animated.timing(opacity, {
        toValue: 0,
        duration: 200,
        useNativeDriver: true,
      }),
      Animated.timing(translateY, {
        toValue: -20,
        duration: 200,
        useNativeDriver: true,
      }),
    ]).start(() => {
      setIsVisible(false);
      opacity.setValue(0);
      translateY.setValue(-20);
      onDismiss?.();
    });
  };

  const handleAction = () => {
    if (dismissTimeoutRef.current) {
      clearTimeout(dismissTimeoutRef.current);
    }
    onAction?.();
    handleDismiss();
  };

  if (!isVisible || !context) return null;

  const bannerInfo = BannerTypeInfo[context.bannerType];

  return (
    <Animated.View
      style={[
        styles.container,
        { backgroundColor: bannerInfo.backgroundColor },
        {
          opacity,
          transform: [{ translateY }],
        },
      ]}
      accessibilityRole="alert"
      accessibilityLabel={`${bannerInfo.accessibilityLabel}. ${context.bannerMessage}`}
    >
      {/* Icon */}
      <Ionicons
        name={bannerInfo.icon}
        size={24}
        color={bannerInfo.iconColor}
        style={styles.icon}
      />

      {/* Content */}
      <View style={styles.content}>
        <Text style={styles.message}>{context.bannerMessage}</Text>

        {/* Action button */}
        {context.hasActionButton && onAction && (
          <TouchableOpacity
            style={[styles.actionButton, { backgroundColor: bannerInfo.iconColor }]}
            onPress={handleAction}
            accessibilityLabel={context.actionButtonTitle}
            accessibilityHint="Opens rest day check-in options"
          >
            <Text style={styles.actionButtonText}>{context.actionButtonTitle}</Text>
          </TouchableOpacity>
        )}
      </View>

      {/* Dismiss button */}
      <TouchableOpacity
        style={styles.dismissButton}
        onPress={handleDismiss}
        accessibilityLabel="Dismiss"
        accessibilityHint="Dismisses this notification banner"
      >
        <Ionicons name="close" size={16} color="#8E8E93" />
      </TouchableOpacity>
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    padding: 16,
    marginHorizontal: 16,
    marginVertical: 8,
    borderRadius: 12,
    borderWidth: 1,
    borderColor: 'rgba(0,0,0,0.05)',
  },
  icon: {
    marginRight: 12,
    marginTop: 2,
  },
  content: {
    flex: 1,
  },
  message: {
    fontSize: 15,
    fontWeight: '500',
    color: '#000',
    lineHeight: 22,
  },
  actionButton: {
    alignSelf: 'flex-start',
    paddingHorizontal: 16,
    paddingVertical: 8,
    borderRadius: 20,
    marginTop: 12,
  },
  actionButtonText: {
    fontSize: 14,
    fontWeight: '600',
    color: '#fff',
  },
  dismissButton: {
    padding: 8,
    backgroundColor: '#F2F2F7',
    borderRadius: 20,
    marginLeft: 8,
  },
});

File: MuscleHamster/Views/Notifications/NotificationContextBanner.swift
//
//  NotificationContextBanner.swift
//  MuscleHamster
//
//  Contextual banner shown when app opens from a notification tap
//  Phase 08.3: Notification Tap Routing and Today Context
//

import SwiftUI

// MARK: - Notification Context Banner

/// Banner shown at top of Home after user taps a notification
struct NotificationContextBanner: View {
    let context: NotificationContext
    let onDismiss: () -> Void
    let onAction: (() -> Void)?

    @State private var isVisible: Bool = false
    @State private var dismissTask: Task<Void, Never>?
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    init(
        context: NotificationContext,
        onDismiss: @escaping () -> Void,
        onAction: (() -> Void)? = nil
    ) {
        self.context = context
        self.onDismiss = onDismiss
        self.onAction = onAction
    }

    var body: some View {
        bannerContent
            .opacity(isVisible ? 1 : 0)
            .offset(y: isVisible ? 0 : -20)
            .onAppear {
                triggerAppear()
            }
            .onDisappear {
                dismissTask?.cancel()
            }
    }

    private var bannerContent: some View {
        HStack(spacing: 12) {
            // Icon
            Image(systemName: context.bannerType.icon)
                .font(.title2)
                .foregroundStyle(context.bannerType.iconColor)
                .accessibilityHidden(true)

            // Message and optional action
            VStack(alignment: .leading, spacing: 6) {
                Text(context.bannerMessage)
                    .font(.subheadline)
                    .fontWeight(.medium)
                    .foregroundStyle(.primary)
                    .fixedSize(horizontal: false, vertical: true)

                // Action button for "action needed" state
                if context.hasActionButton, let onAction = onAction {
                    Button(action: {
                        onAction()
                        dismiss()
                    }) {
                        Text(context.actionButtonTitle)
                            .font(.subheadline)
                            .fontWeight(.semibold)
                            .foregroundStyle(.white)
                            .padding(.horizontal, 12)
                            .padding(.vertical, 6)
                            .background(context.bannerType.iconColor)
                            .clipShape(Capsule())
                    }
                    .accessibilityLabel(context.actionButtonTitle)
                    .accessibilityHint("Opens rest day check-in options")
                }
            }

            Spacer()

            // Dismiss button
            Button(action: dismiss) {
                Image(systemName: "xmark")
                    .font(.caption)
                    .fontWeight(.semibold)
                    .foregroundStyle(.secondary)
                    .padding(8)
                    .background(Color(.systemGray5))
                    .clipShape(Circle())
            }
            .accessibilityLabel("Dismiss")
            .accessibilityHint("Dismisses this notification banner")
        }
        .padding()
        .background(context.bannerType.backgroundColor)
        .clipShape(RoundedRectangle(cornerRadius: 12))
        .overlay(
            RoundedRectangle(cornerRadius: 12)
                .strokeBorder(context.bannerType.iconColor.opacity(0.3), lineWidth: 1)
        )
        .accessibilityElement(children: .combine)
        .accessibilityLabel("\(context.bannerType.accessibilityLabel). \(context.bannerMessage)")
        .accessibilityAddTraits(.isStaticText)
    }

    // MARK: - Lifecycle

    private func triggerAppear() {
        // Animate in
        withAnimation(reduceMotion ? .none : .spring(response: 0.4, dampingFraction: 0.8)) {
            isVisible = true
        }

        // Announce for VoiceOver
        announceForVoiceOver()

        // Schedule auto-dismiss if applicable
        if context.shouldAutoDismiss {
            scheduleAutoDismiss()
        }
    }

    private func dismiss() {
        // Cancel any pending auto-dismiss
        dismissTask?.cancel()

        // Animate out
        withAnimation(reduceMotion ? .none : .easeOut(duration: 0.2)) {
            isVisible = false
        }

        // Notify parent after animation
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.25) {
            onDismiss()
        }
    }

    private func scheduleAutoDismiss() {
        dismissTask = Task {
            try? await Task.sleep(nanoseconds: UInt64(context.autoDismissDelay * 1_000_000_000))

            guard !Task.isCancelled else { return }

            await MainActor.run {
                dismiss()
            }
        }
    }

    private func announceForVoiceOver() {
        // Post accessibility announcement for VoiceOver users
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
            UIAccessibility.post(
                notification: .announcement,
                argument: context.bannerMessage
            )
        }
    }
}

// MARK: - Banner Container

/// Container view that manages the notification banner display
struct NotificationBannerContainer: View {
    @EnvironmentObject private var routingState: AppRoutingState
    let onRestDayAction: () -> Void

    var body: some View {
        if let context = routingState.notificationContext {
            NotificationContextBanner(
                context: context,
                onDismiss: {
                    routingState.clearNotificationContext()
                },
                onAction: context.hasActionButton ? onRestDayAction : nil
            )
            .onAppear {
                // Trigger the banner's appear animation
                // The banner handles its own animation internally
            }
            .padding(.horizontal)
            .padding(.top, 8)
        }
    }
}

// MARK: - Preview

#Preview("Streak Safe") {
    VStack {
        NotificationContextBanner(
            context: NotificationContext(
                notificationType: .dailyReminder,
                tappedAt: Date(),
                hasCheckedInToday: true,
                currentStreak: 7
            ),
            onDismiss: { print("Dismissed") }
        )
        .onAppear {
            // Preview needs to trigger appear
        }

        Spacer()
    }
    .padding()
}

#Preview("Action Needed") {
    VStack {
        NotificationContextBanner(
            context: NotificationContext(
                notificationType: .streakAtRisk,
                tappedAt: Date(),
                hasCheckedInToday: false,
                currentStreak: 5
            ),
            onDismiss: { print("Dismissed") },
            onAction: { print("Action tapped") }
        )

        Spacer()
    }
    .padding()
}

#Preview("Gentle Nudge") {
    VStack {
        NotificationContextBanner(
            context: NotificationContext(
                notificationType: .dailyReminder,
                tappedAt: Date(),
                hasCheckedInToday: false,
                currentStreak: 0
            ),
            onDismiss: { print("Dismissed") }
        )

        Spacer()
    }
    .padding()
}
